File: backend/app/schemas/edits.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field


class EditPayload(BaseModel):
    """Body for both preview and apply."""
    changes: dict[str, Any] = Field(default_factory=dict)


class ChangeImpactResponse(BaseModel):
    column: str
    scope: Literal["user", "customer", "dataset"]
    old_value: Any | None = None
    new_value: Any | None = None
    affected_row_count: int


class PreviewResponse(BaseModel):
    impacts: list[ChangeImpactResponse]


class ApplyResponse(BaseModel):
    impacts: list[ChangeImpactResponse]
    refresh_ok: bool
    grants_ok: bool
    refresh_error: str | None = None
    grants_error: str | None = None


class EditableColumnDescriptor(BaseModel):
    """Surfaced by /api/edit/columns so the frontend knows what's editable
    without duplicating the registry."""
    name: str
    scope: Literal["user", "customer", "dataset"]
    kind: Literal["int", "str", "bigint"]
    nullable: bool
    max_length: int | None = None
    min_value: int | None = None
    max_value: int | None = None
    allowed_values: list[Any] | None = None

File: frontend/src/api/edits.ts
/**
 * Edit API client.
 *
 * Two calls per row edit: preview (dry-run, returns impact counts) and
 * apply (commits the change). The frontend renders the confirmation modal
 * off the preview response so CS agents see the exact impact before
 * hitting Confirm.
 */
import { api } from './client'

export type EditScope = 'user' | 'customer' | 'dataset'

export interface EditableColumnDescriptor {
  name: string
  scope: EditScope
  kind: 'int' | 'str' | 'bigint'
  nullable: boolean
  max_length: number | null
  min_value: number | null
  max_value: number | null
  allowed_values: unknown[] | null
}

export function listEditableColumns(): Promise<EditableColumnDescriptor[]> {
  return api.get<EditableColumnDescriptor[]>('/api/edit/columns')
}

export interface ChangeImpact {
  column: string
  scope: EditScope
  old_value: unknown
  new_value: unknown
  affected_row_count: number
}

export interface PreviewResponse {
  impacts: ChangeImpact[]
}

export interface ApplyResponse {
  impacts: ChangeImpact[]
  refresh_ok: boolean
  grants_ok: boolean
  refresh_error: string | null
  grants_error: string | null
}

export interface EditPayload {
  changes: Record<string, unknown>
}

export function previewChanges(
  userId: string,
  databaseName: string,
  customerCode: number,
  changes: Record<string, unknown>,
): Promise<PreviewResponse> {
  return api.post<PreviewResponse>(
    `/api/users/${encodeURIComponent(userId)}/${encodeURIComponent(
      databaseName,
    )}/preview?customer_code=${customerCode}`,
    { changes },
  )
}

export function applyChanges(
  userId: string,
  databaseName: string,
  customerCode: number,
  changes: Record<string, unknown>,
): Promise<ApplyResponse> {
  return api.patch<ApplyResponse>(
    `/api/users/${encodeURIComponent(userId)}/${encodeURIComponent(
      databaseName,
    )}?customer_code=${customerCode}`,
    { changes },
  )
}
